src/logs_storage/mod.rs

```diff
@@ -1,149 +1,263 @@
+        COMMIT;"#;
+
+    pub const CLEANUP: &str = "DELETE FROM query_logs_v2 WHERE timestamp < ?";
+
+    pub const INSERT: &str =
+        "INSERT INTO query_logs_v2(query_id, timestamp, log_msg) VALUES(?, ?, ?)";
+
+    pub const SELECT: &str = r#"
+        SELECT log_msg, query_id
+        FROM query_logs_v2
+        WHERE
+            timestamp >= :timestamp
+            AND (
+                :last_query_id IS NULL
+                OR rowid > (SELECT rowid FROM query_logs_v2 WHERE query_id = :last_query_id)
+            )
+        ORDER BY rowid
+        "#;
 }
 
-// TODO: reimplement according to https://github.com/subsquid/sqd-network/issues/122
 impl LogsStorage {
     pub async fn new(logs_path: &str) -> Result<Self> {
         let db = Connection::open(logs_path).await?;
-        let state = db
-            .call(|db| {
-                db.execute_batch(
-                    r#"
-                    BEGIN;
-                    CREATE TABLE IF NOT EXISTS query_logs(seq_no INTEGER PRIMARY KEY, log_msg BLOB);
-                    CREATE TABLE IF NOT EXISTS next_seq_no(seq_no);
-                    COMMIT;"#,
-                )?;
-                let has_next_seq_no = db
-                    .prepare_cached("SELECT seq_no FROM next_seq_no")?
-                    .exists(())?;
-                match has_next_seq_no {
-                    false => Ok(InitState::NotInitialized),
-                    true => Ok(InitState::Initialized),
-                }
-            })
-            .await?;
 
-        Ok(Self {
-            db,
-            init_state: state.into(),
+        db.call(|db| {
+            db.execute_batch(sql::CREATE)?;
+            Ok(())
         })
-    }
+        .await?;
 
-    pub fn is_initialized(&self) -> bool {
-        self.init_state.load(Ordering::SeqCst) == InitState::Initialized
+        Ok(Self { db })
     }
 
-    pub async fn save_log(&self, _log: QueryExecuted) -> Result<()> {
-        // assert!(self.is_initialized());
-        // self.db
-        //     .call(move |db| {
-        //         let tx = db.transaction()?;
-        //         log.timestamp_ms = timestamp_now_ms();
-        //         tx.prepare_cached("INSERT INTO query_logs SELECT seq_no, ? FROM next_seq_no")
-        //             .expect("Couldn't prepare logs insertion query")
-        //             .execute([log.encode_to_vec()])?;
-        //         tx.prepare_cached("UPDATE next_seq_no SET seq_no = seq_no + 1")
-        //             .expect("Couldn't prepare next_seq_no update query")
-        //             .execute(())?;
-        //         tx.commit()?;
-        //         Ok(())
-        //     })
-        //     .await?;
+    pub async fn save_log(&self, log: QueryExecuted) -> Result<()> {
+        let query = log.query.as_ref().expect("Log should be well formed");
+        let timestamp = query.timestamp_ms;
+        let id = query.query_id.clone();
+        let log_msg = log.encode_to_vec();
```

Comment:
Yeah, I also thought about it, but then I imagined how painful it would be to update the schema in such case. So let's keep it like this at least until the message schema stabilizes